scripts/centralize_docs.py

```diff
@@ -0,0 +1,180 @@
+
+def get_new_documents(temp_dir_path, pmcids):
+
+    # Save pmids to a new file
+    pmcids_file = temp_dir_path / 'pmcids.txt'
+    pmcids = [str(pmcid) for pmcid in pmcids]
+    pmcids_file.write_text('\n'.join(pmcids))
+
+    # Run pubget run command using subprocess module
+    out_dir = temp_dir_path / 'output'
+    subprocess.run(['pubget', 'download', '--pmcids_file',
+                    str(pmcids_file), str(out_dir)])
+
+    res_dir = list(out_dir.glob('pmcidList*'))[0]
+
+    subprocess.run(['pubget', 'extract_articles', 
+                    str(res_dir / 'articlesets')])
+
+    subprocess.run(['pubget', 'extract_data', 
+                    str(res_dir / 'articles')])
+
+    subprocess.run(['pubget', 'extract_labelbuddy_data',
+                    str(res_dir / 'subset_allArticles_extractedData')])
+
+    json_files = list(
+        (res_dir / 'subset_allArticles_labelbuddyData').glob('*.jsonl'))
+
+    # Load old and new documents
+    new_docs = {}
+    for file in json_files:
+        new_docs.update(_load_jsonl(file))
+
+    return json_files, new_docs
+
+
+def get_old_documents():
+    old_docs = {}
+    all_pmids = set()
+    for file in pathlib.Path('projects').glob('*/documents/*.jsonl'):
+        # Skip symlinked files
+        if file.is_symlink():
+            continue
+        docs = _load_jsonl(file)
+
+        if not docs:
+            # Skip documents that don't have pmcids
+            continue
+
+        old_docs[file] = docs
+        all_pmids.update(docs.keys())
+
+    return old_docs, all_pmids
+
+
+def centralize_all():
+    old_docs, all_old_pmcids = get_old_documents()
+    temp_dir_path = pathlib.Path('/tmp/pubget')
+    json_files, new_docs = get_new_documents(temp_dir_path, all_old_pmcids)
+
+    # Check for document differences
+    diff_docs = defaultdict(dict)
+    for f_name, old_docs_vals in old_docs.items():
+        for pmcid, old_doc_j in old_docs_vals.items():
+            if pmcid in new_docs:
+                old_text = old_doc_j['text']
+                new_text = new_docs[pmcid]['text']
+
+                if old_text != new_text:
+                    diff_docs[pmcid] = (deepcopy(old_doc_j), new_docs[pmcid])
+
+    # If differences, check if annotations are still valid
+    # If not valid, raise an error and don't update
+    annotations = pathlib.Path('projects').glob('*/annotations/*.json*')
+    for a in annotations:
+        if a.suffix == '.json':
+            with open(a, 'r') as f:
+                data = json.loads(f.read())
+        else:
+            data = [json.loads(line) for line in a.open()]
+
+        for d in data:
+            if 'pmid' in d['metadata']:
+                pmcid = d['metadata']['pmid']
+            else:
+                continue
+            if pmcid in diff_docs:
+                if not check_annotations(d['annotations'], diff_docs[pmcid]):
+                    raise ValueError(
+                        "Incompatible annotation found. Can't update documents"
```

Comment:
so I guess this means there are no incompatible annotations in the currently existing ones?

Reply:
Correct.